### PR TITLE
[move-analyzer] Cache user program

### PR DESCRIPTION
**Pull request by awelc**
## Description 

This PR builds upon the compiler changes supporting partial compilation of of user code (skipping function bodies when compiling specific files designated by the user) to deliver the performance improvements described in the [PR](https://github.com/MystenLabs/sui/pull/20588) containing the compiler changes (reducing compilation time in the IDE for the Deepbook package from over 1s to less than 200ms).

In this PR, in addition to caching dependencies, we also cache user code, fully compile only modified files, and merge result of this compilation with the cached data

## Test plan 

All tests must pass. Also tested manually to verify correctness and performance improvements

### CODE REVIEW

File: external-crates/move/crates/move-analyzer/src/symbols.rs [hunk 1]
```diff
@@ -155,11 +155,9 @@ pub struct CompiledPkgInfo {
     /// A combined hash for manifest files of the dependencies
     deps_hash: String,
     /// Information about cached dependencies
-    cached_deps: Option<CachedDeps>,
+    cached_deps: Option<AnalyzedPkgInfo>,
     /// Compiled user program
     program: CompiledProgram,
-    /// Source files
```
Comment (awelc) [inline external-crates/move/crates/move-analyzer/src/symbols.rs:161]: Somewhat unrelated, but I realized we don't need both

File: external-crates/move/crates/move-analyzer/tests/ide_testsuite.rs [hunk 2]
```diff
@@ -381,6 +381,7 @@ fn initial_symbols(
         pkg_deps.clone(),
         ide_files_root.clone(),
         project_path.as_path(),
+        None,
         LintLevel::None,
     )?;
```
Comment (cgswords) [inline external-crates/move/crates/move-analyzer/tests/ide_testsuite.rs:387]: Can we add a test with some precompiled info just for sanity checking?

File: external-crates/move/crates/move-analyzer/src/symbols.rs [hunk 3]
```diff
@@ -1427,17 +1434,45 @@ impl SymbolicatorRunner {
...
                             }
-                            pkgs_to_analyze
-                                .entry(root_dir.unwrap())
-                                .or_insert_with(BTreeSet::new)
-                                .insert(starting_path.clone());
+                            // we want to collect either all modified Move files (as BTreeSet<Some(PathBuf)> or,
+                            // if one of the paths for a package is not a Move file, have aonly one BTreeSet
+                            // None entry. This will be later passed to symbolicator as Some(Vec<PathBuf>) or
+                            // None, to indicate if incremental symbolication is possible or not.
+                            let root_dir = root_dir_opt.unwrap();
```
Comment (cgswords) [inline external-crates/move/crates/move-analyzer/src/symbols.rs:1441]: Small nit: possibly swap this for a let-else

File: external-crates/move/crates/move-analyzer/src/symbols.rs [hunk 4]
```diff
@@ -1406,10 +1413,10 @@ impl SymbolicatorRunner {
                         }
                     };
                     if let Some(all_starting_paths) = all_starting_paths_opt {
```
Comment (cgswords) [inline external-crates/move/crates/move-analyzer/src/symbols.rs:1415]: I am really confused about the naming of the variables in this code here:

- Why `Some(root_dir)` when that is a set?
- What does `all_starting_paths` _mean_?

File: external-crates/move/crates/move-analyzer/src/symbols.rs [hunk 5]
```diff
@@ -1427,17 +1434,45 @@ impl SymbolicatorRunner {
...
-                            pkgs_to_analyze
-                                .entry(root_dir.unwrap())
-                                .or_insert_with(BTreeSet::new)
-                                .insert(starting_path.clone());
+                            // we want to collect either all modified Move files (as BTreeSet<Some(PathBuf)> or,
+                            // if one of the paths for a package is not a Move file, have aonly one BTreeSet
+                            // None entry. This will be later passed to symbolicator as Some(Vec<PathBuf>) or
+                            // None, to indicate if incremental symbolication is possible or not.
+                            let root_dir = root_dir_opt.unwrap();
+                            if let Some(mut modified_files) = pkgs_to_analyze.remove(&root_dir) {
```
Comment (cgswords) [inline external-crates/move/crates/move-analyzer/src/symbols.rs:1442]: Is there a reason to `remove` and `insert` vs simply using `entry` with the appropriate `std::mem::replace`?

File: external-crates/move/crates/move-analyzer/src/symbols.rs [hunk 6]
```diff
@@ -1427,17 +1434,45 @@ impl SymbolicatorRunner {
...
+                                // we already seen this package
+                                if modified_files.len() != 1 || modified_files.first().is_some() {
```
Comment (cgswords) [inline external-crates/move/crates/move-analyzer/src/symbols.rs:1444]: What is this testing for and why not just `!modified_files.is_empty()`?

File: external-crates/move/crates/move-analyzer/src/symbols.rs [hunk 7]
```diff
@@ -1427,17 +1434,45 @@ impl SymbolicatorRunner {
...
+                                    // keep adding modfied Move files if any
+                                    if is_move_file(starting_path) {
+                                        // add move file to existing set
+                                        modified_files.insert(Some(starting_path.clone()));
+                                        pkgs_to_analyze.insert(root_dir, modified_files);
+                                    } else {
+                                        // reset to a single None entry
+                                        pkgs_to_analyze.insert(root_dir,  BTreeSet::from_iter(vec![None]));
```
Comment (cgswords) [inline external-crates/move/crates/move-analyzer/src/symbols.rs:1452]: Can you explain this? If we have a bunch of modified move files and one non-move file, why not at least consider the move files?

File: external-crates/move/crates/move-analyzer/src/symbols.rs [hunk 8]
```diff
@@ -1427,17 +1434,45 @@ impl SymbolicatorRunner {
...
+                                    }
+                                } // else we already seen this package with a non-Move-file path so nothing to do
+                            } else {
+                                // first time we see this package
```
Comment (cgswords) [inline external-crates/move/crates/move-analyzer/src/symbols.rs:1456]: It seems like this logic could be combined with the above logic if we used `entry`

File: external-crates/move/crates/move-analyzer/src/symbols.rs [hunk 9]
```diff
@@ -1763,26 +1798,104 @@ impl Symbols {
...
+    files_to_compile: &BTreeSet<PathBuf>,
+    file_hashes: Arc<BTreeMap<PathBuf, FileHash>>,
+) -> bool {
+    files_to_compile.iter().any(|fpath| {
+        if let Some(fhash) = file_hashes.get(fpath) {
+            if let P::Definition::Module(mdef) = &pkg_def.def {
+                return mdef.loc.file_hash() == *fhash;
+            }
+        }
+        false
```
Comment (cgswords) [inline external-crates/move/crates/move-analyzer/src/symbols.rs:1818]: No strong preference, but could write as:

```
file_hashes.get(fpath)
    .and_then(|fhash| match &pkg_def.def {
        P::Definition::Module(mdef) => Some((mdef, fhash)),
        _ => None,
    })
    .map(|(mdef, fhash)| mdef.loc.file_hash() == *fhash)
    .unwrap_or(false)
    ```

File: external-crates/move/crates/move-analyzer/src/symbols.rs [hunk 10]
```diff
@@ -1763,26 +1798,104 @@ impl Symbols {
...
+fn is_typed_mod_modified(
+    mdef: &ModuleDefinition,
+    files_to_compile: &BTreeSet<PathBuf>,
+    file_hashes: Arc<BTreeMap<PathBuf, FileHash>>,
+) -> bool {
+    files_to_compile.iter().any(|fpath| {
+        if let Some(fhash) = file_hashes.get(fpath) {
+            return mdef.loc.file_hash() == *fhash;
+        }
+        false
```
Comment (cgswords) [inline external-crates/move/crates/move-analyzer/src/symbols.rs:1831]:
```suggestion
        file_hashes.get(fpath)
    .map(|fhash| mdef.loc.file_hash() == *fhash)
    .unwrap_or(false)
```

**Review by cgswords — APPROVED** [review]
LGTM modulo the code in `new` for `SymbolicatorRunner`, which is opaque and very hard to follow. There are several bits that do unexpected things. Refactoring that massive constructor into smaller functions that do clear things would help quite a bit, and make the code that much easier to maintain going forward.

File: external-crates/move/crates/move-analyzer/src/symbols.rs (hunk 3 above)
Comment (awelc) [reply external-crates/move/crates/move-analyzer/src/symbols.rs:1441]: Good idea! Thanks!

File: external-crates/move/crates/move-analyzer/src/symbols.rs (hunk 6 above)
Comment (awelc) [reply external-crates/move/crates/move-analyzer/src/symbols.rs:1444]: It could be `!modified_files.is_empty()`. The intention was to represent a situation when `modified_files` has only one entry and this entry is `None`, but since an entry cannot be `None` if other entries exist, checking of non-emptiness is totally fine as well

File: external-crates/move/crates/move-analyzer/src/symbols.rs (hunk 7 above)
Comment (awelc) [reply external-crates/move/crates/move-analyzer/src/symbols.rs:1452]: On the second look, I overdid it as there is no way that there can be non-Move file paths in this function (though they can exist as input to `get_symbols`). I now simply collect the file paths

File: external-crates/move/crates/move-analyzer/src/symbols.rs (hunk 8 above)
Comment (awelc) [reply external-crates/move/crates/move-analyzer/src/symbols.rs:1456]: Logic changed so no longer needed

File: external-crates/move/crates/move-analyzer/src/symbols.rs (hunk 10 above)
Comment (awelc) [reply external-crates/move/crates/move-analyzer/src/symbols.rs:1831]: Thanks! An arguably even more Rust-y one would be:
```rust
file_hashes
  .get(fpath)
  .map_or(false, |fhash| mdef.loc.file_hash() == *fhash)
```

**Comment by awelc** [comment]
> LGTM modulo the code in `new` for `SymbolicatorRunner`, which is opaque and very hard to follow. There are several bits that do unexpected things. Refactoring that massive constructor into smaller functions that do clear things would help quite a bit, and make the code that much easier to maintain going forward.

Moved aggregation of package paths into a separate function which was the part that seemed to make the most sense to refactor

File: external-crates/move/crates/move-analyzer/tests/ide_testsuite.rs (hunk 2 above)
Comment (awelc) [reply external-crates/move/crates/move-analyzer/tests/ide_testsuite.rs:387]: We need a full compilation to compute all symbols and this is the first and only one that's done per test suite (as files do not change). What I did, though, is to add another partial compilation right after this one and use its results to run tests, but it lengthens the time to execute the test suite (not to unreasonable level, though)
